File: data_structure/tree/binary_tree/binary_tree.go/binarytree/binarytree.go
package binarytree

import (
	"fmt"
)

type Node struct {
	lchild, rchild *Node
	value          int
}

type Tree struct {
	root *Node
}

func (t *Tree) AddElement(v int) {
	if t == nil {
		return
	}

	n := t.root
	var p *Node
	for n != nil {
		p = n
		if n.value < v {
			n = n.rchild
		} else if v < n.value {
			n = n.lchild
		} else {
			return
		}
	}

	if p == nil {
		t.root = &Node{value: v}
	} else {
		if p.value < v {
			p.rchild = &Node{value: v}
		} else {
			p.lchild = &Node{value: v}
		}
	}
}

func (t *Tree) RemoveElement(v int) {
	if t == nil {
		return
	}

	n := t.root
	var p *Node
	for n != nil {
		p = n
		if n.value < v {
			n = n.rchild
		} else if v < n.value {
			n = n.lchild
		} else {
			break
		}
	}

	if n == nil {
		return
	}

	var c *Node
	var cp *Node

	if n.rchild != nil {
		cp = n
		c = n.rchild
		for c.lchild != nil {
			cp = c
			c = c.lchild
		}
	} else if n.lchild != nil {
		cp = n
		c = n.lchild
		for c.rchild != nil {
			cp = c
			c = c.rchild
		}
	}

	if c == nil && cp == nil {
		if p.value < n.value {
			p.rchild = nil
		} else {
			p.lchild = nil
		}
	} else {
		var cc *Node
		if c.lchild != nil {
			cc = c.lchild
		} else if c.rchild != nil {
			cc = c.rchild
		}

		if c.value < cp.value {
			cp.lchild = cc
		} else {
			cp.rchild = cc
		}

		n.value = c.value
	}
}

func (t *Tree) SearchElement(v int) bool {
	if t == nil {
		return false
	}

	n := t.root
	for n != nil {
		if n.value < v {
			n = n.rchild
		} else if v < n.value {
			n = n.lchild
		} else {
			return true
		}
	}

	return false
}

func (t *Tree) FreeTree() {
	if t != nil {
		t.root = nil
	}
}

// traverse tree

func traversePreorder(n *Node) {
	if n == nil {
		return
	}

	fmt.Printf("%v ", n.value)
	traversePreorder(n.lchild)
	traversePreorder(n.rchild)
}

func (t *Tree) TraversePreorder() {
	if t == nil {
		return
	}

	traversePreorder(t.root)
}

func traverseInorder(n *Node) {
	if n == nil {
		return
	}

	traverseInorder(n.lchild)
	fmt.Printf("%v ", n.value)
	traverseInorder(n.rchild)
}

func (t *Tree) TraverseInorder() {
	if t == nil {
		return
	}

	traverseInorder(t.root)
}

func traversePostorder(n *Node) {
	if n == nil {
		return
	}

	traversePostorder(n.lchild)
	traversePostorder(n.rchild)
	fmt.Printf("%v ", n.value)
}

func (t *Tree) TraversePostorder() {
	if t == nil {
		return
	}

	traversePostorder(t.root)
}

File: data_structure/tree/binary_tree/binary_tree.cpp/binarytree.cpp


#include "binarytree.h"


#include <iostream>



void binarytree::Tree::Add(int k)
{
	Node *n = root;
	Node *p = n;

	while (n != nullptr) {
		p = n;

		if (n->k < k) {
			n = n->rchild;
		}
		else if (k < n->k) {
			n = n->lchild;
		}
		else {
			return;
		}
	}

	if (nullptr == p) {
		root = GetNode();
		root->k = k;
	}
	else {
		if (p->k < k) {
			p->rchild = GetNode();
			p->rchild->k = k;
		}
		else {
			p->lchild = GetNode();
			p->lchild->k = k;
		}
	}
}

void binarytree::Tree::Remove(int k)
{
	Node *n = root;
	Node *p = n;
	while (n != nullptr)
	{
		if (n->k < k) {
			p = n;
			n = n->rchild;
		}
		else if (k < n->k) {
			p = n;
			n = n->lchild;
		}
		else {
			break;
		}
	}

	if (nullptr == n)
		return;

	Node *replaceone = nullptr;
	Node *replaceonep = nullptr;
	if (n->lchild != nullptr) {
		replaceone = n->lchild;
		while (replaceone->rchild != nullptr) {
			replaceonep = replaceone;
			replaceone = replaceone->rchild;
		}

		if (replaceonep != nullptr) {
			replaceonep->rchild = replaceone->lchild;
		}
		else {
			n->lchild = replaceone->lchild;
		}
		n->k = replaceone->k;
	}
	else if (n->rchild != nullptr) {
		replaceone = n->rchild;
		while (replaceone->lchild != nullptr)
		{
			replaceonep = replaceone;
			replaceone = replaceonep->lchild;
		}

		if (replaceonep != nullptr) {
			replaceonep->lchild = replaceonep->rchild;
		}
		else {
			n->rchild = replaceone->rchild;
		}
		n->k = replaceone->k;
	}
	else {
		// root value
		if (p == n) {
			DesctroyNode(root);
			root = nullptr;
		}
		else {
			if (p->k < n->k) {
				p->rchild = nullptr;
			}
			else {
				p->lchild = nullptr;
			}

			DesctroyNode(n);
		}
	}
}

bool binarytree::Tree::Search(int k)
{
	Node *n = root;
	while (n != nullptr)
	{
		if (n->k < k) {
			n = n->rchild;
		}
		else if (k < n->k) {
			n = n->lchild;
		}
		else {
			return true;
		}
	}

	return false;
}

void binarytree::Tree::TraversePreorder()
{
	TraverseNodePreorder(root);
}

void binarytree::Tree::TraverseInorder()
{
	TraverseNodeInorder(root);
}

void binarytree::Tree::TraversePostorder()
{
	TraverseNodePostorder(root);
}

void binarytree::Tree::TraverseNodePreorder(Node * n)
{
	if (nullptr == n)
		return;
	std::cout << n->k << " ";
	TraverseNodePreorder(n->lchild);
	TraverseNodePreorder(n->rchild);

}

void binarytree::Tree::TraverseNodeInorder(Node * n)
{
	if (nullptr == n)
		return;
	TraverseNodeInorder(n->lchild);
	std::cout << n->k << " ";
	TraverseNodeInorder(n->rchild);
}

void binarytree::Tree::TraverseNodePostorder(Node * n)
{
	if (nullptr == n)
		return;
	TraverseNodePostorder(n->lchild);
	TraverseNodePostorder(n->rchild);
	std::cout << n->k << " ";
}

binarytree::Tree::Node* binarytree::Tree::GetNode()
{
	return new Node();
}

void binarytree::Tree::DesctroyNode(Node * n)
{
	if (n != nullptr) {
		delete n;
	}
}
